Mathlib/CategoryTheory/Abelian/DiagramLemmas/Horseshoe.lean

```diff
@@ -0,0 +1,685 @@
+      horseshoe_X, π₂_obj, HomologicalComplex.shortComplexFunctor_obj_X₂,
+      HomologicalComplex.shortComplexFunctor_obj_X₃, HomologicalComplex.shortComplexFunctor_obj_f,
+      Functor.mapHomologicalComplex_obj_d, horseshoe_d, ChainComplex.prev, ↓reduceDite, π₂_map,
+      comp_τ₂, HomologicalComplex.shortComplexFunctor_obj_g, ChainComplex.next_nat_succ,
+      exact_iso_comp]
+
+    set g := _; change CategoryTheory.Exact _ g
+    dsimp at g
+    suffices g_eq : g = (horseshoeD A n).τ₂ ≫ eqToHom (by simp) by
+      rw [g_eq, exact_comp_iso]
+      exact this
+    simp only [g]
+    generalize_proofs h1 h2
+    rw [show (eqToHom h1).τ₂ = eqToHom (by simp) from stupid_aux' _ _ (by simp), stupid_aux2'']
+    simp
+
+  suffices HomologicalComplex.homologyMap (horseshoeToSingle₂ A) (n + 1) = (z2.iso z1).hom by
+    rw [this]
+    exact IsIso.of_iso _
+  exact IsZero.eq_of_tgt z1 _ _
+
+instance (n : ℕ) : QuasiIsoAt (horseshoeToSingle₃ A) (n + 1) := by
+  rw [quasiIsoAt_iff_isIso_homologyMap]
+  have z1 : IsZero <| ((ChainComplex.single₀ _).obj A.X₃).homology (n + 1) := by
+    apply HomologicalComplex.isZero_single_obj_homology
+    simp
+
+  have z2 : IsZero <| HomologicalComplex.homology (horseshoeChainComplex₃ A) (n + 1) := by
+    suffices e : HomologicalComplex.homology (horseshoeChainComplex₃ A) (n + 1) ≅ 0 from
+      e.isZero_iff.mpr (isZero_zero _)
+
+    refine exact_iff_homology_iso_zero _ |>.mp ?_ |>.some
+    rw [← exact_iff_shortComplex_exact]
+    have := horseshoeExact₃ A n
+    simp only [Functor.mapHomologicalComplex_obj_X, horseshoe_X, π₂_obj,
+      Functor.mapHomologicalComplex_obj_d, horseshoe_d, eqToHom_refl, Category.id_comp, dite_eq_ite,
+      π₂_map, ↓reduceDite] at this
+    rw [if_pos (by abel)] at this
+
+    simp only [HomologicalComplex.shortComplexFunctor_obj_X₁, Functor.mapHomologicalComplex_obj_X,
+      horseshoe_X, π₃_obj, HomologicalComplex.shortComplexFunctor_obj_X₂,
+      HomologicalComplex.shortComplexFunctor_obj_X₃, HomologicalComplex.shortComplexFunctor_obj_f,
+      Functor.mapHomologicalComplex_obj_d, horseshoe_d, ChainComplex.prev, ↓reduceDite, π₃_map,
+      comp_τ₃, HomologicalComplex.shortComplexFunctor_obj_g, ChainComplex.next_nat_succ,
+      exact_iso_comp]
+
+    set g := _; change CategoryTheory.Exact _ g
+    dsimp at g
+    suffices g_eq : g = (horseshoeD A n).τ₃ ≫ eqToHom (by simp) by
+      rw [g_eq, exact_comp_iso]
+      exact this
+    simp only [g]
+    generalize_proofs h1 h2
+    rw [show (eqToHom h1).τ₃ = eqToHom (by simp) from stupid_aux'' _ _ (by simp), stupid_aux2''']
+    simp
+
+  suffices HomologicalComplex.homologyMap (horseshoeToSingle₃ A) (n + 1) = (z2.iso z1).hom by
+    rw [this]
+    exact IsIso.of_iso _
+  exact IsZero.eq_of_tgt z1 _ _
+
+instance : _root_.QuasiIso (horseshoeToSingle₁ A) where
+  quasiIsoAt n := by
+    cases n <;> infer_instance
+
+instance : _root_.QuasiIso (horseshoeToSingle₂ A) where
```

Comment:
All of this is great! As we already have a construction of a projective resolution, would not it be easier to prove the horseshoe lemma as follows, given a short exact sequence `0 ⟶ X₁ ⟶ X₂ ⟶ X₃ ⟶ 0` and projective resolution of `X₁` and `X₃`, construct a projective resolution of `X₂` which fits in between? This may be slightly easier than reconstructing everything. The only data which would have to be produced is the differential on the "naive" coproduct of the two resolutions.

Reply:
I just tried to define the differential $P_{n + 1} \oplus Q_{n + 1} \to P_n \oplus Q_n$ where $P$ is a projective resolution for $X_1$ and $Q$ for $X_3$ but couldn't get it to work, one of the commutativity I can't prove. I documented the failed attempt at `Horseshoe'.lean.failed_attempt`

Reply:
Feel free to re-use the draft code at https://github.com/leanprover-community/mathlib4/compare/master...horseshoe-jriou where I construct the relevant part of the middle differential `δ`, the middle complex is `Qcomplex`, and it fits into a short complex `shortComplex` of chain complexes, and I have defined the augmentation `shortComplexε`, so that I think all the compatibilities should be ok. Then, we have to show that `shortComplex` is short exact, which is easy because degreewise we have a splitting. Finally, more importantly, in order to show that the `Qcomplex` is a (projective) resolution, you may use lemmas which are not yet in #12638 but shall soon be PRed (see https://github.com/leanprover-community/mathlib4/blob/jriou_localization/Mathlib/Algebra/Homology/HomologySequenceLemmas.lean#L270 )